### PR TITLE
Add shared element transitions

### PR DESCRIPTION
**Pull request by FooIbar**

### CODE REVIEW

**Comment by revonateB0T** [comment]
https://issuetracker.google.com/issues/336449371

**Review by FooIbar — COMMENTED** [review]
https://issuetracker.google.com/333675164

File: app/src/main/kotlin/com/hippo/ehviewer/ui/tools/Modifier.kt [hunk 1]
```diff
@@ -1,6 +1,39 @@
...
     if (condition) block() else this
+
+val NoopTransitionsVisibilityScope = TransitionsVisibilityScope(emptySet())
+
+@Stable
+class TransitionsVisibilityScope(val scopes: Set<AnimatedVisibilityScope>)
+
+context(TransitionsVisibilityScope)
+@Composable
+inline fun <T> AnimatedVisibilityScope.advance(block: @Composable TransitionsVisibilityScope.() -> T) = block(remember(scopes, this) { TransitionsVisibilityScope(scopes + this) })
```
Comment (revonateB0T) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/tools/Modifier.kt:24]: The `advance` should have a better name

**Comment by revonateB0T** [comment]
Wow, seems both SET within navigations and downloadScreen grid/list switch works perfectly!

**Comment by revonateB0T** [comment]
Btw, my telegram account gone again, and maybe I won't register a new one in a long time.

File: app/src/main/kotlin/com/hippo/ehviewer/ui/tools/Modifier.kt [hunk 2]
```diff
@@ -1,6 +1,39 @@
...
+import androidx.compose.runtime.remember
 import androidx.compose.ui.Modifier
 
 inline fun Modifier.thenIf(condition: Boolean, crossinline block: Modifier.() -> Modifier) =
     if (condition) block() else this
+
+val NoopTransitionsVisibilityScope = TransitionsVisibilityScope(emptySet())
+
+@Stable
+class TransitionsVisibilityScope(val scopes: Set<AnimatedVisibilityScope>)
```
Comment (revonateB0T) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/tools/Modifier.kt:20]: Can we make it value class for better performance?

File: app/src/main/kotlin/com/hippo/ehviewer/ui/ComposeDefault.kt [hunk 3]
```diff
@@ -1,25 +1,36 @@
...
-    LocalContext.current.run { remember { findActivity<MainActivity>() } },
-    LocalSnackBarHostState.current,
-    LocalDialogState.current,
-    navigator,
-    rememberCoroutineScope(),
-)
+) = with(NoopTransitionsVisibilityScope) {
+    advance {
+        block(
+            LocalContext.current.run { remember { findActivity<MainActivity>() } },
```
Comment (revonateB0T) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/ComposeDefault.kt:27]: Better to be `with(LocalContext.current) { findActivity() } `
`MainActivity` can be inferred I think.

**Comment by revonateB0T** [comment]
Is there anything blocking we land this?

**Comment by FooIbar** [comment]
> Is there anything blocking we land this?

Haven't got time to look at this

File: app/src/main/kotlin/com/hippo/ehviewer/ui/tools/Modifier.kt (hunk 1 above)
Comment (FooIbar) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/tools/Modifier.kt:25]: Move `AnimatedVisibilityScope` to parameter and name it `with`?

Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/tools/Modifier.kt:25]: SGTM.

**Comment by FooIbar** [comment]
Everything looks good, seems we don't need https://issuetracker.google.com/issues/336449371?

File: app/src/main/kotlin/com/hippo/ehviewer/ui/ComposeDefault.kt [hunk 4]
```diff
@@ -22,7 +22,7 @@ inline fun <R> AnimatedVisibilityScope.composing(
     block: @Composable context(MainActivity, SnackbarHostState, DialogState, SharedTransitionScope, TransitionsVisibilityScope, DestinationsNavigator, CoroutineScope)
     () -> R,
 ) = with(NoopTransitionsVisibilityScope) {
-    advance {
+    togetherWith(this@AnimatedVisibilityScope) {
```
Comment (revonateB0T) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/ComposeDefault.kt:25]: Use `implicit` instead.

File: app/src/main/kotlin/com/hippo/ehviewer/ui/screen/DownloadsScreen.kt [hunk 5]
```diff
@@ -535,7 +535,7 @@ fun AnimatedVisibilityScope.DownloadsScreen(navigator: DestinationsNavigator) =
             navToReader(info.galleryInfo)
         }
         AnimatedContent(gridView, transitionSpec = { fadeIn() togetherWith fadeOut() }, label = "Downloads") { showGridView ->
-            advance {
+            togetherWith(this@AnimatedVisibilityScope) {
```
Comment (revonateB0T) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/screen/DownloadsScreen.kt:538]: Ditto.

**Comment by revonateB0T** [comment]
> Everything looks good, seems we don't need https://issuetracker.google.com/issues/336449371?

Yeah, could you help test if that snippet still behaves incorrectly?

**Review by revonateB0T — APPROVED** [review]

**Review by revonateB0T — APPROVED** [review]

**Comment by FooIbar** [comment]
> > Everything looks good, seems we don't need https://issuetracker.google.com/issues/336449371?
> 
> Yeah, could you help test if that snippet still behaves incorrectly?

Yes, the issue persists.

**Comment by revonateB0T** [comment]
> > > Everything looks good, seems we don't need https://issuetracker.google.com/issues/336449371?
> > 
> > 
> > Yeah, could you help test if that snippet still behaves incorrectly?
> 
> Yes, the issue persists.

Acked.

**Comment by FooIbar** [comment]
@revonateB0T 
I think we may only want shared element transitions for list <=> detail.
The list <=> list and list <=> grid transitions feel distracting as there are too many moving parts.
Like the M3 guide says https://m3.material.io/styles/motion/transitions/applying-transitions#b974d94c-bf80-4c54-8818-4a1d7eaa10ae

**Comment by revonateB0T** [comment]
> @revonateB0T I think we may only want shared element transitions for list <=> detail. The list <=> list and list <=> grid transitions feel distracting as there are too many moving parts. Like the M3 guide says https://m3.material.io/styles/motion/transitions/applying-transitions#b974d94c-bf80-4c54-8818-4a1d7eaa10ae

Well, maybe, list <=> grid is somewhat lag as there's so many moving items.
But I suppose the multiple SET framework can be kept in case it's useful and we may need it one day.

**Pull request merged**